File: agents/sql_generator.py
"""
AskERP — SQL Generator Agent (Day 9)

Takes a user question + retrieved schema/metric chunks and produces
executable DuckDB SQL via Claude claude-sonnet-4-6.

Importable: from agents.sql_generator import generate_sql
CLI:        python3 -m agents.sql_generator "What was our total revenue in 2024?"
             (fetches context from retrieval API on localhost:8001)
"""

import os
import sys
import json
import re

BASE_DIR = os.path.join(os.path.dirname(__file__), "..")

# Load API key from .env.local if not already in environment
def _load_env():
    env_path = os.path.join(BASE_DIR, ".env.local")
    if os.path.exists(env_path):
        with open(env_path) as f:
            for line in f:
                line = line.strip()
                if line and not line.startswith("#") and "=" in line:
                    key, _, val = line.partition("=")
                    if key.strip() not in os.environ:
                        os.environ[key.strip()] = val.strip().strip('"').strip("'")

_load_env()

try:
    import anthropic
except ImportError:
    print("ERROR: anthropic not installed. Run: pip3 install anthropic")
    sys.exit(1)

_CLIENT = anthropic.Anthropic(api_key=os.environ.get("ANTHROPIC_API_KEY"))

# ── System prompt ──────────────────────────────────────────────────────────────
_SYSTEM_PROMPT = """You are a SQL generator for a DuckDB analytics warehouse for Northwind Furniture, a ~$50M B2B furniture company.

DATABASE FACTS:
- DuckDB dialect (similar to PostgreSQL/BigQuery)
- Fiscal year: calendar year (Jan-Dec)
- Date dimension: dim_date with columns: date_id (YYYYMMDD integer), date (DATE), year, quarter, month
- All fact tables join to dim_date via date_id foreign keys
- fact_sales_order has: order_id, customer_id, employee_id, order_date_id, location_id, order_status, total_amount, line_count, primary_item_category, effective_unit_cost, effective_gross_margin_pct
- order_status values: 'Confirmed', 'Delivered', 'Cancelled', 'Pending'
- Active orders = order_status NOT IN ('Cancelled', 'Pending')

RULES:
1. Return ONLY valid DuckDB SQL — no CTEs unless necessary, no window functions unless asked
2. Always filter fact_sales_order to exclude Cancelled/Pending unless user asks for cancellations
3. Use dim_date for all date filtering — do not filter on date_id integers directly
4. Aggregate to reasonable granularity — if user says "by quarter" add dd.year, dd.quarter to GROUP BY
5. LIMIT results to 100 rows maximum
6. Never use SELECT * — name the columns explicitly
7. Use table aliases: fso=fact_sales_order, dc=dim_customer, di=dim_item, dd=dim_date, de=dim_employee, dp=fact_payment, fi=fact_invoice

GROSS MARGIN FORMULA:
(SUM(fso.total_amount) - SUM(fso.effective_unit_cost * fso.line_count)) / SUM(fso.total_amount) * 100

Output JSON with exactly these keys:
{
  "sql": "SELECT ...",
  "metric_used": "metric_name or null",
  "sentiment": "positive or negative or none",
  "tables_referenced": ["table1", "table2"],
  "reasoning": "one sentence",
  "confidence": "HIGH or MEDIUM or LOW"
}

For sentiment: if a metric is used, derive it from the metric's sentiment field in the retrieved context.
Use "positive" for metrics where higher is better (revenue, margin, active customers),
"negative" for metrics where higher is worse (churn, cancellation, days to pay),
"none" for neutral or directional metrics (growth rate, segment breakdown).
If no metric is used, set "none".

Output ONLY the JSON object. No markdown, no explanation."""

# ── Few-shot examples ──────────────────────────────────────────────────────────
_FEW_SHOT_EXAMPLES = [
    {
        "question": "What was our total revenue in 2024?",
        "answer": json.dumps({
            "sql": "SELECT SUM(fso.total_amount) AS total_revenue FROM fact_sales_order fso JOIN dim_date dd ON fso.order_date_id = dd.date_id WHERE fso.order_status NOT IN ('Cancelled', 'Pending') AND dd.year = 2024",
            "metric_used": "total_revenue",
            "sentiment": "positive",
            "tables_referenced": ["fact_sales_order", "dim_date"],
            "reasoning": "Sum total_amount for confirmed/delivered orders in 2024 using dim_date year filter.",
            "confidence": "HIGH"
        })
    },
    {
        "question": "Top 5 customers by revenue last quarter",
        "answer": json.dumps({
            "sql": "SELECT dc.customer_name, SUM(fso.total_amount) AS revenue FROM fact_sales_order fso JOIN dim_customer dc ON fso.customer_id = dc.customer_id JOIN dim_date dd ON fso.order_date_id = dd.date_id WHERE fso.order_status NOT IN ('Cancelled', 'Pending') AND dd.year = 2025 AND dd.quarter = 4 GROUP BY dc.customer_name ORDER BY revenue DESC LIMIT 5",
            "metric_used": "total_revenue",
            "sentiment": "positive",
            "tables_referenced": ["fact_sales_order", "dim_customer", "dim_date"],
            "reasoning": "Join customer dimension, group by name, order by revenue descending, limit 5.",
            "confidence": "HIGH"
        })
    },
    {
        "question": "Average gross margin for Outdoor Furniture in 2025",
        "answer": json.dumps({
            "sql": "SELECT ROUND((SUM(fso.total_amount) - SUM(fso.effective_unit_cost * fso.line_count)) / SUM(fso.total_amount) * 100, 2) AS gross_margin_pct FROM fact_sales_order fso JOIN dim_date dd ON fso.order_date_id = dd.date_id WHERE fso.order_status NOT IN ('Cancelled', 'Pending') AND fso.primary_item_category = 'Outdoor Furniture' AND dd.year = 2025",
            "metric_used": "gross_margin_pct",
            "sentiment": "positive",
            "tables_referenced": ["fact_sales_order", "dim_date"],
            "reasoning": "Filter to Outdoor Furniture using primary_item_category, apply gross margin formula.",
            "confidence": "HIGH"
        })
    },
    {
        "question": "Cancellation rate by quarter in 2024",
        "answer": json.dumps({
            "sql": "SELECT dd.quarter, ROUND(COUNT(CASE WHEN fso.order_status = 'Cancelled' THEN 1 END) * 100.0 / COUNT(*), 2) AS cancellation_rate_pct FROM fact_sales_order fso JOIN dim_date dd ON fso.order_date_id = dd.date_id WHERE dd.year = 2024 GROUP BY dd.quarter ORDER BY dd.quarter",
            "metric_used": "cancellation_rate",
            "sentiment": "negative",
            "tables_referenced": ["fact_sales_order", "dim_date"],
            "reasoning": "Count cancelled orders as pct of all orders, grouped by quarter.",
            "confidence": "HIGH"
        })
    },
    {
        "question": "Which customers churned in late 2024?",
        "answer": json.dumps({
            "sql": "SELECT dc.customer_name, MAX(dd.date) AS last_order_date FROM fact_sales_order fso JOIN dim_customer dc ON fso.customer_id = dc.customer_id JOIN dim_date dd ON fso.order_date_id = dd.date_id GROUP BY dc.customer_name HAVING MAX(dd.date) < '2024-11-01' AND MAX(dd.date) > '2024-06-01' ORDER BY last_order_date DESC LIMIT 20",
            "metric_used": "customer_churn_count",
            "sentiment": "negative",
            "tables_referenced": ["fact_sales_order", "dim_customer", "dim_date"],
            "reasoning": "Find customers whose last order was between mid-2024 and Nov 2024, indicating churn.",
            "confidence": "MEDIUM"
        })
    },
]


def _build_user_message(question: str, retrieved_chunks: list) -> str:
    parts = ["RETRIEVED CONTEXT:\n"]

    for chunk in retrieved_chunks[:6]:  # cap at 6 chunks to stay within context
        parts.append(f"[{chunk['id']}]")
        parts.append(chunk.get("text", "")[:600])
        parts.append("")

    parts.append("\nFEW-SHOT EXAMPLES:")
    for ex in _FEW_SHOT_EXAMPLES:
        parts.append(f'Q: {ex["question"]}')
        parts.append(f'A: {ex["answer"]}')
        parts.append("")

    parts.append(f"\nQUESTION: {question}")
    parts.append("\nGenerate the SQL now. Output only the JSON object.")

    return "\n".join(parts)


def generate_sql(question: str, retrieved_chunks: list) -> dict:
    """
    Given a user question and retrieved schema/metric chunks, generate executable SQL.

    Returns:
    {
      "sql": "SELECT ...",
      "metric_used": str | None,
      "tables_referenced": list[str],
      "reasoning": str,
      "confidence": "HIGH" | "MEDIUM" | "LOW"
    }
    """
    user_msg = _build_user_message(question, retrieved_chunks)

    response = _CLIENT.messages.create(
        model="claude-sonnet-4-6",
        max_tokens=1024,
        system=_SYSTEM_PROMPT,
        messages=[{"role": "user", "content": user_msg}],
    )

    raw = response.content[0].text.strip()

    # Strip markdown code fences if present
    raw = re.sub(r"^```(?:json)?\s*", "", raw)
    raw = re.sub(r"\s*```$", "", raw)

    try:
        result = json.loads(raw)
    except json.JSONDecodeError as e:
        return {
            "sql": None,
            "metric_used": None,
            "tables_referenced": [],
            "reasoning": f"JSON parse error: {e}. Raw: {raw[:200]}",
            "confidence": "LOW",
            "error": "json_parse_failed",
        }

    return result


# ── CLI mode ───────────────────────────────────────────────────────────────────
if __name__ == "__main__":
    import urllib.request

    if len(sys.argv) < 2:
        print("Usage: python3 -m agents.sql_generator \"your question here\"")
        sys.exit(1)

    question = sys.argv[1]
    print(f'\nQuestion: "{question}"')
    print("Fetching context from retrieval API...")

    # Call retrieval service
    try:
        payload = json.dumps({"query": question, "k": 5}).encode()
        req = urllib.request.Request(
            "http://localhost:8001/retrieve",
            data=payload,
            headers={"Content-Type": "application/json"},
            method="POST",
        )
        with urllib.request.urlopen(req, timeout=10) as resp:
            retrieval_response = json.loads(resp.read())
        chunks = retrieval_response.get("results", [])
        print(f"Retrieved {len(chunks)} chunks (confidence: {retrieval_response.get('confidence', {}).get('confidence_band', '?')})")
    except Exception as e:
        print(f"ERROR: Could not reach retrieval API at localhost:8001 — {e}")
        print("Start it with: uvicorn retriever.api_server:app --port 8001")
        sys.exit(1)

    print("Generating SQL...")
    result = generate_sql(question, chunks)
    print("\nResult:")
    print(json.dumps(result, indent=2))

File: lib/sqlGenerator.ts
import Anthropic from "@anthropic-ai/sdk";
import fs from "fs";
import path from "path";
import * as yaml from "js-yaml";
import type { RetrievalResult } from "./retrieval/retriever";

const client = new Anthropic({ apiKey: process.env.ANTHROPIC_API_KEY });

// USE_RETRIEVAL=true → prompt built from retrieved chunks only; false → full schema in prompt (A/B flag)
export const USE_RETRIEVAL = process.env.USE_RETRIEVAL === "true";

// Tenant this route runs against (D-035/D-037). Mirrors the retriever's own
// ACTIVE_TENANT default so both layers agree without a shared module.
export const DEFAULT_TENANT = process.env.ACTIVE_TENANT ?? "mro";

const ROOT = process.cwd();
const schemaYaml = fs.readFileSync(path.join(ROOT, "data", "schema.yaml"), "utf-8");
const metricsYaml = fs.readFileSync(path.join(ROOT, "data", "metrics.yaml"), "utf-8");

// data/schema.yaml and data/metrics.yaml hold BOTH tenants in one file
// (tenant: northwind | mro per-entry, D-033). Filter before building any
// full-corpus prompt so a tenant never sees the other tenant's tables.
interface YamlTable { name: string; tenant?: string; [k: string]: unknown }
interface YamlMetric { name: string; tenant?: string; [k: string]: unknown }
const parsedSchema = yaml.load(schemaYaml) as { tables: YamlTable[] };
const parsedMetrics = yaml.load(metricsYaml) as { metrics: YamlMetric[] };

function tenantSchemaYaml(tenant: string): string {
  return yaml.dump({ tables: parsedSchema.tables.filter((t) => t.tenant === tenant) });
}
function tenantMetricsYaml(tenant: string): string {
  return yaml.dump({ metrics: parsedMetrics.metrics.filter((m) => m.tenant === tenant) });
}

const SYSTEM_BASE_NORTHWIND = `You are a SQL generator for a DuckDB analytics warehouse for Northwind Furniture, a ~$50M B2B furniture company.`;

const SYSTEM_RULES_NORTHWIND = `
RULES:
1. Return ONLY valid DuckDB SQL — no CTEs unless necessary
2. Always filter fact_sales_order to exclude Cancelled/Pending unless the user asks about cancellations
3. Use dim_date for all date filtering — do not filter on date_id integers directly
4. dim_date columns: date_id (YYYYMMDD int), date (DATE), year, quarter, month
5. If user asks "by quarter", include both year and quarter in GROUP BY and SELECT
6. LIMIT results to 100 rows maximum
7. Never use SELECT * — name columns explicitly
8. Table aliases: fso=fact_sales_order, dc=dim_customer, di=dim_item, dd=dim_date, de=dim_employee, dp=fact_payment, fi=fact_invoice

GROSS MARGIN FORMULA:
(SUM(fso.total_amount) - SUM(fso.effective_unit_cost * fso.line_count)) / SUM(fso.total_amount) * 100

For sentiment: use the metric's sentiment field from the SEMANTIC LAYER above.
"positive" = higher is better, "negative" = higher is worse, "none" = directional/neutral.

Output ONLY this JSON object — no markdown, no explanation:
{
  "sql": "SELECT ...",
  "metric_used": "metric_name or null",
  "sentiment": "positive | negative | none",
  "tables_referenced": ["table1"],
  "reasoning": "one sentence",
  "confidence": "HIGH | MEDIUM | LOW"
}`;

// MRO Distributor tenant (D-035/D-037). All tables live under schema
// "mro_distributor" inside the MotherDuck database "askerp" (also present in
// the local file at data/northwind.db). System/rules mirror the benchmark's
// proven prompt (D-036: dense retrieval + this shape scored best on H-tier).
const SYSTEM_BASE_MRO = `You are a SQL generator for a DuckDB analytics warehouse for the MRO Distributor tenant (an industrial supplies distributor). All warehouse tables live in the schema "mro_distributor" — always qualify table names as mro_distributor.<table>.`;

const SYSTEM_RULES_MRO = `
RULES:
1. Return ONLY valid DuckDB SQL — no CTEs unless necessary
2. Always qualify every table as mro_distributor.<table_name>
3. Use mro_distributor.dim_date for all date filtering, joined on the *_date_key integer (YYYYMMDD) columns — never filter on the integer key directly
4. The data covers 18 months of history; unless the question names a period, aggregate over all data
5. LIMIT results to 100 rows maximum
6. Never use SELECT * — name columns explicitly
7. If the question cannot be answered from the provided context and this warehouse's data (missing table, out-of-scope topic, data that does not exist here), set "sql" to null and explain why in "reasoning" — do not invent tables or columns

For sentiment: use the metric's sentiment field from the SEMANTIC LAYER above.
"positive" = higher is better, "negative" = higher is worse, "none" = directional/neutral.

Output ONLY this JSON object — no markdown, no explanation:
{
  "sql": "SELECT ...",
  "metric_used": "metric_name or null",
  "sentiment": "positive | negative | none",
  "tables_referenced": ["mro_distributor.table1"],
  "reasoning": "one sentence",
  "confidence": "HIGH | MEDIUM | LOW"
}`;

function tenantBase(tenant: string): string {
  return tenant === "mro" ? SYSTEM_BASE_MRO : SYSTEM_BASE_NORTHWIND;
}
function tenantRules(tenant: string): string {
  return tenant === "mro" ? SYSTEM_RULES_MRO : SYSTEM_RULES_NORTHWIND;
}
function tenantFewShot(tenant: string): string {
  return tenant === "mro" ? FEW_SHOT_MRO : FEW_SHOT_NORTHWIND;
}

// Full-schema prompt (USE_RETRIEVAL=false), filtered to one tenant's tables/metrics.
const fullSystemCache = new Map<string, string>();
function buildFullSystem(tenant: string): string {
  const cached = fullSystemCache.get(tenant);
  if (cached) return cached;
  const built = `${tenantBase(tenant)}

DATABASE SCHEMA:
${tenantSchemaYaml(tenant)}

SEMANTIC LAYER (metrics with formulas):
${tenantMetricsYaml(tenant)}
${tenantRules(tenant)}`;
  fullSystemCache.set(tenant, built);
  return built;
}

function buildRetrievedSystem(chunks: RetrievalResult[], tenant: string): string {
  const tables = chunks.filter((c) => c.type === "table");
  const metrics = chunks.filter((c) => c.type === "metric");
  const tableSection = tables.length > 0
    ? `RETRIEVED SCHEMA CONTEXT (top ${tables.length} relevant tables):\n${tables.map((c) => c.text).join("\n\n")}`
    : "";
  const metricSection = metrics.length > 0
    ? `RETRIEVED METRIC CONTEXT (top ${metrics.length} relevant metrics):\n${metrics.map((c) => c.text).join("\n\n")}`
    : "";
  return [tenantBase(tenant), tableSection, metricSection, tenantRules(tenant)].filter(Boolean).join("\n\n");
}

const FEW_SHOT_MRO = `FEW-SHOT EXAMPLES:

Q: What is our total revenue?
A: {"sql":"SELECT ROUND(SUM(ext_amount),2) AS total_revenue FROM mro_distributor.o2c_sales_order_line","metric_used":"total_revenue","sentiment":"positive","tables_referenced":["mro_distributor.o2c_sales_order_line"],"reasoning":"Sum extended amount across all order lines.","confidence":"HIGH"}

Q: What is our current DSO?
A: {"sql":"SELECT ROUND(SUM(days_to_pay*amount_applied)/SUM(amount_applied),2) AS dso_days FROM mro_distributor.ar_payment_application","metric_used":"dso_days","sentiment":"negative","tables_referenced":["mro_distributor.ar_payment_application"],"reasoning":"Weighted-average days to pay across applied payments.","confidence":"HIGH"}

Q: Which item category has the best gross margin?
A: {"sql":"SELECT i.category, ROUND(SUM(sol.margin_amount)/SUM(sol.ext_amount)*100,2) AS gm_pct FROM mro_distributor.o2c_sales_order_line sol JOIN mro_distributor.dim_item i USING (item_key) GROUP BY 1 ORDER BY 2 DESC","metric_used":"gross_margin_pct","sentiment":"positive","tables_referenced":["mro_distributor.o2c_sales_order_line","mro_distributor.dim_item"],"reasoning":"Margin amount over extended amount, grouped by item category.","confidence":"HIGH"}

Q: Who are our top 10 suppliers by spend?
A: {"sql":"SELECT s.supplier_name, ROUND(SUM(po.ext_cost),2) AS spend FROM mro_distributor.p2p_purchase_order_line po JOIN mro_distributor.dim_supplier s USING (supplier_key) GROUP BY 1 ORDER BY 2 DESC LIMIT 10","metric_used":"spend_by_supplier","sentiment":"none","tables_referenced":["mro_distributor.p2p_purchase_order_line","mro_distributor.dim_supplier"],"reasoning":"Sum extended cost per supplier, top 10 by spend.","confidence":"HIGH"}

Q: What was Northwind Furniture's total revenue?
A: {"sql":null,"metric_used":null,"sentiment":"none","tables_referenced":[],"reasoning":"Northwind is a different tenant; this warehouse only contains MRO Distributor data.","confidence":"HIGH"}`;

const FEW_SHOT_NORTHWIND = `FEW-SHOT EXAMPLES:

Q: What was our total revenue in 2024?
A: {"sql":"SELECT SUM(fso.total_amount) AS total_revenue FROM fact_sales_order fso JOIN dim_date dd ON fso.order_date_id = dd.date_id WHERE fso.order_status NOT IN ('Cancelled','Pending') AND dd.year = 2024","metric_used":"total_revenue","sentiment":"positive","tables_referenced":["fact_sales_order","dim_date"],"reasoning":"Sum active orders in 2024.","confidence":"HIGH"}

Q: Top 5 customers by revenue last quarter
A: {"sql":"SELECT dc.customer_name, SUM(fso.total_amount) AS revenue FROM fact_sales_order fso JOIN dim_customer dc ON fso.customer_id = dc.customer_id JOIN dim_date dd ON fso.order_date_id = dd.date_id WHERE fso.order_status NOT IN ('Cancelled','Pending') AND dd.year = 2025 AND dd.quarter = 4 GROUP BY dc.customer_name ORDER BY revenue DESC LIMIT 5","metric_used":"total_revenue","sentiment":"positive","tables_referenced":["fact_sales_order","dim_customer","dim_date"],"reasoning":"Join customers, group by name, top 5 by revenue.","confidence":"HIGH"}

Q: Cancellation rate by quarter in 2024
A: {"sql":"SELECT dd.year, dd.quarter, ROUND(COUNT(CASE WHEN fso.order_status = 'Cancelled' THEN 1 END) * 100.0 / COUNT(*), 2) AS cancellation_rate_pct FROM fact_sales_order fso JOIN dim_date dd ON fso.order_date_id = dd.date_id WHERE dd.year = 2024 GROUP BY dd.year, dd.quarter ORDER BY dd.year, dd.quarter","metric_used":"cancellation_rate","sentiment":"negative","tables_referenced":["fact_sales_order","dim_date"],"reasoning":"Count cancelled as pct of all orders, grouped by quarter.","confidence":"HIGH"}

Q: Which customers churned in late 2024?
A: {"sql":"SELECT dc.customer_name, MAX(dd.date) AS last_order_date FROM fact_sales_order fso JOIN dim_customer dc ON fso.customer_id = dc.customer_id JOIN dim_date dd ON fso.order_date_id = dd.date_id GROUP BY dc.customer_name HAVING MAX(dd.date) < '2024-11-01' AND MAX(dd.date) > '2024-06-01' ORDER BY last_order_date DESC LIMIT 20","metric_used":"customer_churn_count","sentiment":"negative","tables_referenced":["fact_sales_order","dim_customer","dim_date"],"reasoning":"Last order before Nov 2024 signals churn.","confidence":"MEDIUM"}

Q: Outdoor furniture gross margin in 2025
A: {"sql":"SELECT ROUND((SUM(fso.total_amount) - SUM(fso.effective_unit_cost * fso.line_count)) / SUM(fso.total_amount) * 100, 2) AS gross_margin_pct FROM fact_sales_order fso JOIN dim_date dd ON fso.order_date_id = dd.date_id WHERE fso.order_status NOT IN ('Cancelled','Pending') AND fso.primary_item_category = 'Outdoor Furniture' AND dd.year = 2025","metric_used":"gross_margin_pct","sentiment":"positive","tables_referenced":["fact_sales_order","dim_date"],"reasoning":"Gross margin formula filtered to Outdoor Furniture 2025.","confidence":"HIGH"}`;

export interface SQLResult {
  sql: string | null;
  metric_used: string | null;
  sentiment: string | null;
  tables_referenced: string[];
  reasoning: string;
  confidence: string;
  error?: string;
}

export async function generateSQL(
  question: string,
  retrievedChunks?: RetrievalResult[],
  tenant: string = DEFAULT_TENANT
): Promise<SQLResult> {
  const system =
    USE_RETRIEVAL && retrievedChunks && retrievedChunks.length > 0
      ? buildRetrievedSystem(retrievedChunks, tenant)
      : buildFullSystem(tenant);

  const userMsg = `${tenantFewShot(tenant)}\n\nQUESTION: ${question}\n\nGenerate the SQL now. Output only the JSON object.`;

  const msg = await client.messages.create({
    model: "claude-haiku-4-5-20251001",
    max_tokens: 1024,
    system,
    messages: [{ role: "user", content: userMsg }],
  });

  const raw = (msg.content[0] as { text: string }).text
    .trim()
    .replace(/^```(?:json)?\s*/m, "")
    .replace(/\s*```$/m, "");

  try {
    return JSON.parse(raw) as SQLResult;
  } catch {
    return {
      sql: null,
      metric_used: null,
      sentiment: null,
      tables_referenced: [],
      reasoning: `JSON parse failed. Raw: ${raw.slice(0, 200)}`,
      confidence: "LOW",
      error: "json_parse_failed",
    };
  }
}
